tooi/api/streaming.py

```diff
@@ -0,0 +1,241 @@
+import asyncio
+import json
+import logging
+import re
+
+from httpx import RequestError
+
+from tooi.asyncio import run_async_task
+from tooi.context import get_context
+from tooi.data.instance import InstanceInfo
+
+# Streaming (push) support for timelines.  Streaming is handled by the InstanceStreamer class, which
+# should only be created once per instance.  InstanceStreamer handles connecting to the streaming
+# endpoint, receiving events, and distributing them to the various subscribers.  To subscribe to
+# streaming, use subscribe() to create a StreamSubscription instance.
+
+
+class StreamEvent(object):
+    """Internal type representing an event received from a stream."""
+    def __init__(self, stream: str, event: str, payload):
+        self.stream = stream
+        self.event = event
+        self.payload = payload
+
+
+StreamQueue = asyncio.Queue[StreamEvent]
+logger = logging.getLogger(__name__)
+
+
+class HTTPStreamClient(object):
+    """
+    Client for connecting to a streaming endpoint and fetching events.
+    This implements the HTTP (non-WebSocket) protocol.
+    """
+
+    TIMEOUT = 120
+    ERROR_BACKOFF = 30
+
+    def __init__(
+            self,
+            instance: InstanceInfo,
+            stream_name: str):
+
+        self.instance = instance
+        # TODO: This should be part of the instance.
+        self.ctx = get_context()
+        self.stream_name = stream_name
+        self.url = f"/api/v1/streaming/{self.stream_name}"
+        self.queue = StreamQueue()
+        self._lines = []
+
+    async def close(self):
+        pass
+
+    async def run(self):
+        logger.info(f"HTTPStreamClient: running url={self.url}")
+
+        # Run forever, silently reconnecting if we get an error.
+        while True:
+            try:
+                await self._stream()
+            except RequestError as exc:
+                logger.info((
+                    f"HTTPStreamClient: disconnected from stream={self.stream_name}: "
+                    f"{str(exc)}"))
+                await asyncio.sleep(self.ERROR_BACKOFF)
+
+    async def _stream(self):
+        logger.info("HTTPStreamClient: connecting to stream={stream}")
```

Comment:
```suggestion
        logger.info(f"HTTPStreamClient: connecting to stream={self.stream_name}")
```

Reply:
fixed in 6025178a3b3ce26093985013d8b9f3d6bc86e824